astropy/units/quantity_helper/function_helpers.py

```diff
@@ -0,0 +1,648 @@
+SUBCLASS_SAFE_FUNCTIONS |= {
+    np.isclose, np.allclose,
+    np.array2string, np.array_repr, np.array_str}
+
+# TODO: could be supported but need work & thought.
+UNSUPPORTED_FUNCTIONS |= {
+    np.histogramdd, np.piecewise,
+    np.apply_along_axis, np.apply_over_axes}
+
+# Nonsensical for quantities.
+UNSUPPORTED_FUNCTIONS |= {
+    np.packbits, np.unpackbits, np.unravel_index,
+    np.ravel_multi_index, np.ix_, np.cov, np.corrcoef,
+    np.busday_count, np.busday_offset, np.datetime_as_string,
+    np.is_busday}
+
+# The following are not just unsupported, but so unlikely to be thought
+# to be supported that we ignore them in testing.  (Kept in a separate
+# variable so that we can check consistency in the test routine -
+# test_quantity_non_ufuncs.py)
+IGNORED_FUNCTIONS = {
+    # Deprecated
+    np.rank, np.asscalar,
+    # I/O - useless for Quantity, since no way to store the unit.
+    np.save, np.savez, np.savetxt, np.savez_compressed,
+    # Polynomials
+    np.poly, np.polyadd, np.polyder, np.polydiv, np.polyfit, np.polyint,
+    np.polymul, np.polysub, np.polyval, np.roots, np.vander,
+    # setops
+    np.in1d, np.intersect1d, np.isin, np.setdiff1d,
+    np.setxor1d, np.union1d, np.unique,
+    # financial
+    np.fv, np.ipmt, np.irr, np.mirr, np.nper, np.npv, np.pmt, np.ppmt,
+    np.pv, np.rate}
+UNSUPPORTED_FUNCTIONS |= IGNORED_FUNCTIONS
+
+
+class FunctionAssigner:
+    def __init__(self, assignments):
+        self.assignments = assignments
+
+    def __call__(self, f=None, helps=None):
+        """Add a helper to a numpy function.
+
+        Normally used as a decorator.
+
+        If ``helps`` is given, it should be the numpy function helped (or an
+        iterable of numpy functions helped).
+
+        If ``helps`` is not given, it is assumed the function helped is the
+        numpy function with the same name as the decorated function.
+        """
+        if f is not None:
+            if helps is None:
+                helps = getattr(np, f.__name__)
+            if not isiterable(helps):
+                helps = (helps,)
+            for h in helps:
+                self.assignments[h] = f
+            return f
+        elif helps is not None:
+            return functools.partial(self.__call__, helps=helps)
+        else:  # pragma: no cover
+            raise ValueError("function_helper requires at least one argument.")
+
+
+function_helper = FunctionAssigner(FUNCTION_HELPERS)
+
+dispatched_function = FunctionAssigner(DISPATCHED_FUNCTIONS)
+
+
+@function_helper(helps={np.copy, np.asfarray, np.real_if_close,
+                        np.sort_complex, np.resize})
+def invariant_a_helper(a, *args, **kwargs):
+    return (a.view(np.ndarray),) + args, kwargs, a.unit, None
+
+
+@function_helper(helps={np.tril, np.triu})
+def invariant_m_helper(m, *args, **kwargs):
+    return (m.view(np.ndarray),) + args, kwargs, m.unit, None
+
+
+# Not really logical to allow ones_like (what is the point of it?),
```

Comment:
I'm still confused about this comment - doesn't it make sense to support this? 
```python
>>> np.ones_like([5., 6., 0.]*u.km)
[1., 1., 1.] km
```
makes sense to me!

Reply:
Or maybe I'm misunderstanding something here. Why is `zeros_like()` subclass safe, but `ones_like` and `full_like` are not?

Reply:
Basically, as soon as one covers `empty_like`, one can fill with `0` (since no unit is required), but not with anything else. So, for me, it feels like I create this empty thing with units of `km` and then filling it with `1` is a bit strange. 

On the other hand, often we think of the unit as equivalent to the quantity with just a value of `1` - and this is clearly how you interpret it above. As this is completely fair, I'll replace the comment.